### ISSUE TITLE
Fields in Component config should be optional (pointers)

### ISSUE THREAD

**Issue by mvladev** [kubernetes/kubernetes#78109]
I want to embed [LeaderElectionConfiguration](https://github.com/kubernetes/component-base/blob/4727f38490bc68c3ba5e308586e7b045e216b991/config/v1alpha1/types.go#L27) in my configuration like this:

```go
import (
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	componetv1alpha1 "k8s.io/component-base/config/v1alpha1"
)

type MyComponentConfiguration struct {
	metav1.TypeMeta `json:",inline"`
	// ClientConnection specifies the kubeconfig file and client connection
	// settings for my controller to use when communicating with the apiserver.
	ClientConnection componetv1alpha1.ClientConnectionConfiguration `json:"clientConnection"`
}
```

For my component, I would like to set a *default* value for `ClientConnectionConfiguration.ResourceLock` to `configmaps` and I want to report an error, if end-user provides a wrong configuration.

With the current setup, where `ClientConnectionConfiguration.ResourceLock` is of type string literal (`string`) and without `omitempty` json tag, if the end-user enters invalid configuration - empty string for `resourceLock`, in the defaulting phase I cannot distinguish between invalid configuration (`resourceLock: ""`) or configuration field which is not provided and must be defaulted

```go
if len(obj.ResourceLock) == 0 {
    obj.ResourceLock = "configmaps"
}
```

For this to work, `ClientConnectionConfiguration.ResourceLock` needs to have `omitempty` json tag and be a pointer:

```go
ResourceLock *string `json:"resourceLock,omitempty"`
```

With this setup, If end-user configures my controller with the following:

```yaml
apiVersion: config.mycomponent/v1alpha1
kind: MyComponentConfiguration
clientConnection:
  leaseDuration: 10s
```

I'll simply check in the Defauling phase for 

```go
if obj.ResourceLock == nil {
    obj.ResourceLock = "configmaps"
}
```

and in my validation 

```go
if len(cc.ResourceLock) == 0 {
    allErrs = append(allErrs, field.Invalid(fldPath.Child("resourceLock"), cc.RenewDeadline, "resourceLock is required"))
}
```

And if an user provides an invalid configuration - it'll be correctly reported:

```yaml
apiVersion: config.mycomponent/v1alpha1
kind: MyComponentConfiguration
clientConnection:
  leaseDuration: 10s
  resourceLock: "" # it should not be empty string and this is invalid for my controller 
```

Changing all fields in `k8s.io/component-base/config/v1alpha1` to pointers and add `omitempty` json tag should be the correct way to go forward?

P.S.
Those fields can still be literals for the internal resources `k8s.io/component-base/config`

cc @luxas 

/kind api-change

**Comment by stealthybox** [comment]
/remove-lifecycle rotten

This is a really important.
It's going to take significant effort to implement, and there have been some recorded discussions on Code-gen alternatives.

**Comment by stealthybox** [comment]
Related KEP
https://github.com/kubernetes/enhancements/pull/1173

**Comment by conwaychriscosmo** [comment]
Hi, 

After reading the issue description and some of the comments I want to confirm I am on the right path for the changes I am considering making.  Does it make sense to go through each *.go file in the component-base/config/v1alpha1 directory and replace each field with a pointer and add omitempty to the json definition as suggested?

For example, should I change,
```
type LeaderElectionConfiguration struct {
    LeaderElect *bool 'json:"leaderElect"'
    LeaseDuration metav1.Duration 'json:"leaseDuration"'
...
}
```
to
```
type LeaderElectionConfiguration struct {
    LeaderElect *bool 'json:"leaderElect,omitempty"'
    LeaseDuration *metav1.Duration 'json:"leaseDuration,omitempty"'
...
}
```
and then repeat that pattern of changes?

**Comment by luxas** [comment]
This is not fixed, reopening.

**Comment by luxas** [comment]
It's fixed for the scheduler (#83619), but not for controller-manager, the kubelet or proxy AFAIK.
cc @sttts @mtaufen @stealthybox
